cirq/protocols/qid_shape_protocol.py

```diff
@@ -0,0 +1,168 @@
+# Copyright 2019 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import TYPE_CHECKING, Any, Tuple, TypeVar, Union
+
+from typing_extensions import Protocol
+
+from cirq.type_workarounds import NotImplementedType
+
+if TYPE_CHECKING:
+    # pylint: disable=unused-import
+    import cirq
+
+# This is a special indicator value used by the methods to determine whether or
+# not the caller provided a 'default' argument. It must be of type
+# Tuple[int, ...] to ensure the method has the correct type signature in that
+# case. It is checked for using `is`, so it won't have a false positive if the
+# user provides a different (0,) value.
+RaiseTypeErrorIfNotProvided = (0,)  # type: Any
+# Equal integers outside the range [-5, 256] aren't identically equal with `is`.
+RaiseTypeErrorIfNotProvidedInt = -2**512  # type: Any
+
+TDefault = TypeVar('TDefault')
+
+
+class SupportsQidShape(Protocol):
+    """A unitary, channel, mixture or other object that operates on a known
+    number qubits/qudits/qids, each with a specific dimension."""
+
+    def _qid_shape_(self) -> Union[Tuple[int, ...], NotImplementedType]:
+        """A tuple specifying the dimension of each qid that this object
+        operates on, e.g. (2, 2) for a two-qubit gate.
```

Comment:
All the 2s make this an ambiguous example. (2, 2, 2) for a three qubit gate would be less ambiguous.